[TACHYON-424] S3 client v2

calvinjia opened the pull request:
Rebased version of #891 

https://tachyon.atlassian.net/projects/TACHYON/issues/TACHYON-424

This PR implements an S3 Client using the jets3t library.

A new s3 submodule under /underfs contains the logic for the client. The user may enable this client as their s3 client by removing the s3n:// prefix from the hadoop prefixes in their configuration.

The implementation conforms to our UnderFileSystem interface and unit tests. It also passes all applicable core tests when using the S3UnderFileSystemCluster. Since the S3UnderFileSystem is implemented as a real S3 backend, the tests are disabled by default.

calvinjia commented:
@apc999 Thanks for the feedback, I've updated the code accordingly.

aarondav commented:
Very high level point, but out of curiosity, why was jets3t used as the S3 library over the AWS SDK? There was a recent effort in Hadoop to switch from the s3n jets3t-backed FS to "s3a" which is AWS SDK-backed, which claims several benefits over the prior ([HADOOP-10400](https://issues.apache.org/jira/browse/HADOOP-10400)).

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3152/
Test PASSed.

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystemFactory.java:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  private boolean addAndCheckAWSCredentials(TachyonConf tachyonConf) {
+    String accessKeyConf = "fs.s3n.awsAccessKeyId";
```
An alternative way to define the inline string constants accessKeyConf, it could be easier to maintain the code if we have a static class string member that holds this constant.

Similar comment on secreteKeyConf.

apc999 replied:
Actually, just realized that both const strings "fs.s3n.awsSecretAccessKey" and "fs.s3n.awsAccessKeyId" appear multiple times in this package. It make better sense to have them either in tachyon.Constants, or define them in something like tachyon.underfs.s3.Constants.java, and refer them using the constant variable name rather than using inline strings. In that way, mis-spelling in the string content will be caught by compiler/IDE before runtime.

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystemFactory.java:
```diff
@@ -0,0 +1,66 @@
+
+  @Override
+  public UnderFileSystem create(String path, TachyonConf tachyonConf, Object conf) {
+    Preconditions.checkArgument(path != null, "path may not be null");
```
NIT: checkNotNull also?
NIT2: shall we also check-faill if tachyonConf and conf are null?

apc999 replied:
Actually, since 3rd parameter conf is not really used in this constructor, considering naming this param as unusedConf.

apc999 commented on underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderStorageCluster.java:
```diff
@@ -0,0 +1,65 @@
+
+  private final String awsAccessKey = "";
+  private final String awsSecretKey = "";
+  private final String baseDirectory = ""; // s3n://my-test-bucket/tachyon-test
```
I am confused, baseDirectory seems never updated after initialized as an empty string.

calvinjia replied:
The user should update this value if they wish to run s3 integration tests.

apc999 replied:
Probably leverage conf to set this value?

calvinjia replied:
Oh yeah, good call.

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3OutputStream.java:
```diff
@@ -0,0 +1,80 @@
+  private final S3Service mClient;
+
+  public S3OutputStream(String bucketName, String key, S3Service client) throws IOException {
+    mBucketName = bucketName;
```
NIT: is bucketName eligible to be empty? if no, probably have a checkArgument here?

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,520 @@
+  // Same as create(path)
+  @Override
+  public OutputStream create(String path, int blockSizeByte) throws IOException {
+    return create(path);
```
Considering LOG.error (or LOG.warn?) here to indicate blockSizeByte is not used, before return create(path).

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,520 @@
+  public boolean delete(String path, boolean recursive) throws IOException {
+    if (!recursive) {
+      if (isFolder(path) && listInternal(path, false).length != 0) {
+        return false;
```
Better to LOG.error indicating why delete fails before return false

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,520 @@
+        return false;
+      }
+    }
+    deleteInternal(path);
```
return deleteInternal(path) ?

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,520 @@
+  // Not supported
+  @Override
+  public Object getConf() {
+    return null;
```
Maybe LOG.error("not supported")?

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,520 @@
+  @Override
+  public boolean rename(String src, String dst) throws IOException {
+    if (!exists(src)) {
+      return false;
```
better to have LOG.error before return false.

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,520 @@
+            + "exists as a file.");
+        return false;
+      }
+      String srcName = getKeyName(src);
```
I am confused here. When isFolder(dst), why we should proceed? 
It seems here we copy src under folder dst in this case? Not sure this matches the semantics of rename in hdfs/gluster ufs.

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,520 @@
+  }
+
+  /**
+   * Strips the s3 bucket prefix from the path if it is present
```
Consider add an example here in javadoc.

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,520 @@
+   * @param key the key to strip
+   * @return the key without the s3 bucket prefix
+   */
+  private String stripPrefix(String key) {
```
how about stripPrefixIfPresent, to better indicate its function?

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,520 @@
+  }
+
+  /**
+   * Determines if the key is represents a folder. If false is returned, it is not guaranteed that
```
delete either "is" or "represents"

apc999 commented:
Made a bunch of comments. Most are tiny readability issues, except:
(1) no real unit-tests on S3UnderFileSystems?
(2) I think the semantics of rename is different from HDFS ufs.  should we be consistent?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3157/
Test PASSed.

calvinjia commented:
@aarondav Thanks for the info, you bring up a very good point. Originally I thought to use AWS SDK, but I found that the jets3t library / s3n was more widely used and mature. When s3a gains more adoption, it should not be a problem to create another underfs for it; this s3 client is by no means the only option that will exist.

@apc999 Thanks for the review, I've updated the code accordingly.
For the first point, the s3 module will pull in all the tests from core (including the under storage system interface test), which is the same test coverage we currently have for the hdfs. However it is disabled by default because the user will need to use an actual S3 backend. 
For the second point, I've changed it to not support rename to existing folders. The updated code is much simpler!

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3159/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3158/
Test PASSed.

haoyuan commented on underfs/s3/pom.xml:
```diff
@@ -0,0 +1,96 @@
+    <dependency>
+      <groupId>net.java.dev.jets3t</groupId>
+      <artifactId>jets3t</artifactId>
+      <version>0.9.3</version>
```
should `jets3t ` be excluded from `hadoop-client` dependency?

calvinjia replied:
The direct dependency will take precedence over the transitive dependency, so it shouldn't be necessary.

haoyuan commented on underfs/s3/pom.xml:
```diff
@@ -0,0 +1,96 @@
+    <dependency>
+      <groupId>org.apache.thrift</groupId>
+      <artifactId>libthrift</artifactId>
+      <scope>test</scope>
```
why is this needed if this references org.tachyonproject.tachyon?

calvinjia replied:
This was previously included because thrift was shaded in tachyon project, but I think it can be removed now since we changed our shading. Good catch.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3160/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3161/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3162/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3165/
Test PASSed.

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,486 @@
+    return isRoot(path) || getObjectDetails(path) != null;
+  }
+
+  // Largest size of one file is 5 TB
```
Could you comment in javadoc why 5TB? is this a requirement from S3? or a limitation in Tachyon?

haoyuan replied:
It's great to change this to be java doc

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,486 @@
+   * @return the key without the s3 bucket prefix
+   */
+  private String stripPrefixIfPresent(String key) {
+    String prefix = Constants.HEADER_S3N + mBucketName + PATH_SEPARATOR;
```
NIT: this prefix really seems a constant after constructor is called, maybe worth have a private member string to prevent string calucation each time on stripPrefixIfPresent

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,486 @@
+      Set<String> children = new HashSet<String>();
+      for (S3Object obj : objs) {
+        // Remove parent portion of the key
+        String child = obj.getKey().substring(path.length());
```
This logic to derive child name could be a separate function in this class, and I highly recommend to unit test this function. Maybe not for this PR though.

apc999 commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,486 @@
+   * @param key the key to get the parent of
+   * @return the the parent key
+   */
+  private String getParentKey(String key) {
```
helper functions like this one and the next should be easily unit-tested, without requiring a S3 mock. Maybe add some unittests (in addition to integration tests) in next PR after unit-tests and integration tests are separated (TACHYON-432)

apc999 commented:
Added a few minor issues.

LGTM :)

Thanks for the PR. this is really an important feature

haoyuan commented on underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java:
```diff
@@ -0,0 +1,486 @@
+ */
+public class S3UnderFileSystem extends UnderFileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private static final String FOLDER_SUFFIX = "_$folder$";
```
Great to add more comments
